### ISSUE TITLE
Offset between slider labels and positions 

### ISSUE THREAD

**Issue by kupiqu** [jspsych/jsPsych#695]
The offset can be seen moving the slider for instance when imposing allowed slider positions to [-5, -4, -3, -2, -1, 0, +1, +2, +3, +4, +5] and setting these values as labels.

**Comment by jodeleeuw** [comment]
![slider](https://user-images.githubusercontent.com/595524/80223541-692a3e00-8616-11ea-978e-c02cbe20cc55.png)

I agree, it looks like the slider width is slightly too narrow and things don't line up quite right. Will require some tweaks to the CSS in the slider plugins.

**Comment by kupiqu** [comment]
Yes, and I think it's a bit more obvious if using `+` in e.g. `+4`. Sorry I forgot to add those in the original post.

**Comment by becky-gilbert** [comment]
I think the offset between thumb position and label comes from the fact that the center of the slider thumb doesn't go all the way to each end of the slider. Instead, the thumb stops when one edge reaches the end of the slider (see below), so the distance between the min and max thumb points is slightly less than than 100%.

![jspsych-slider-label-offset](https://user-images.githubusercontent.com/9041788/97253158-f79b9a00-17c8-11eb-9c43-83286bef117e.png)

We could easily adjust the label positioning by subtracting the thumb width from the slider width, and then base the label positions on that. Except that I can't figure out what the default thumb width is, and I don't know whether it changes across browsers etc. 

One option is to specify the thumb width so that we can adjust for it, but this requires hiding the default slider/thumb and recreating it in CSS (not as bad as it sounds - examples [here](https://www.w3schools.com/howto/howto_js_rangeslider.asp)). @jodeleeuw are you happy with that solution? Other suggestions?

**Comment by jodeleeuw** [comment]
I did a little bit of investigation here. If the labels on the left side are `text-align: left` and positioned at `left: 0`, `left: 10%`, etc. and the labels on the right side are `text-align: right` and positioned at `right: 0`, `right: 10%`, etc. then it seems to align relatively nicely. It's a bit messier to write the plugin code, but maybe avoids custom range CSS?

**Comment by jodeleeuw** [comment]
Actually, this isn't a good solution. It depends on the labels being short, which I hadn't considered. If the label text is more than a few characters it won't work.

**Comment by jodeleeuw** [comment]
Oof. I feel like there is no great option here. I'm leaning towards adding some custom CSS. We already provide custom CSS for buttons via the .jspsych-btn class. Maybe we should just style range sliders as well -- in some minimal fashion that gives us control over this.

**Comment by becky-gilbert** [comment]
Agreed about the lack of good options here. I'll put together some simple custom range slider/thumb styling, and make sure that this solves the problem of the label/thumb alignment, then submit a pull request for review.

**Comment by gorkang** [comment]
This is probably another instance of the same problem, but just in case I wanted to let you know.

Using a slightly modified version of the example shown in the documentation (https://www.jspsych.org/plugins/jspsych-html-slider-response/)

```
var question1 = {
    type: 'html-slider-response',
    stimulus: '<p>Running</p>',
    min: 0,
    max: 100,
    slider_width: 500,
    labels: ['0', '50', '100'],
    prompt: "<p>How healthy or unhealthy this activity is?</p>"
};
```
The alignment of the 50 and 100 anchors is way off. The same happens with the original text anchors used in the documentation, but the effect is a bit more clear using numbers. 

![Screenshot from 2020-11-10 16-50-56](https://user-images.githubusercontent.com/6491790/98705116-445ca480-2375-11eb-9da0-355047e78ce7.png)

**Comment by becky-gilbert** [comment]
Yikes. Thanks for letting us know! We're planning to fix this in the next release.
What browser did you use in the screenshot? I see that you get the slider value on the right - I'm guessing that's the default behavior for the browser? Or is that something you added?

**Comment by gorkang** [comment]
Thanks @becky-gilbert ! I am using Chrome on Linux. 

I believe the slider value is on the right because of the `slider_width: 500,` parameter (I actually like it more on the right, so it is fine for me).

Cheers.

**Comment by jodeleeuw** [comment]
Is the blue 50 something that you added, or is that showing up on the page?

**Comment by gorkang** [comment]
I am so sorry! Yes, it is something we modified, and it is definitely behind the issue in this case. Using the jsPsych-master branch it looks almost perfect. 

![Screenshot from 2020-11-10 20-57-10](https://user-images.githubusercontent.com/6491790/98732679-5f401080-2397-11eb-8797-c550d80bb298.png)

**Comment by becky-gilbert** [comment]
Got it, thanks @gorkang!

**Comment by becky-gilbert** [comment]
How's this?

**Chrome**
![chrome_slider](https://user-images.githubusercontent.com/9041788/98885031-257a1300-2446-11eb-97f4-535ed92fc722.png)

**Firefox**
![firefox_slider](https://user-images.githubusercontent.com/9041788/98885050-3165d500-2446-11eb-9633-90927ab6c409.PNG)

**Edge**
![edge_slider](https://user-images.githubusercontent.com/9041788/98885063-388ce300-2446-11eb-86fc-0180cce599bc.PNG)

And with longer labels:

**Chrome**
![chrome_slider_long](https://user-images.githubusercontent.com/9041788/98885077-42164b00-2446-11eb-941d-31abdf6bed5e.PNG)

**Firefox**
![firefox_slider_long](https://user-images.githubusercontent.com/9041788/98885101-4cd0e000-2446-11eb-8513-cb3d489f44f8.PNG)

**Edge**
![edge_slider_long](https://user-images.githubusercontent.com/9041788/98885134-5c502900-2446-11eb-8a82-aa11ca41571b.PNG)

**Comment by jodeleeuw** [comment]
@becky-gilbert that's a great start! can I play with the CSS a bit? is it in a fork somewhere?

**Comment by becky-gilbert** [comment]
@jodeleeuw yes please do! I just pushed it to the slider-positions branch on my fork:
https://github.com/becky-gilbert/jsPsych/tree/slider-positions
The slider/thumb CSS is in the jspsych.css file. And there are changes to the JavaScript that controls the label positioning, but so far I've only made these changes in the [html-slider-response plugin](https://github.com/becky-gilbert/jsPsych/blob/slider-positions/plugins/jspsych-html-slider-response.js), so you can test using that example file. 
Right now the label positioning requires a known slider width in pixels, so it will break if the slider width isn't specified in the trial parameters. I was waiting to see if you're happy with the changes/approach before fixing that.

**Comment by jodeleeuw** [comment]
I somehow messed up my attempt to PR into your branch and don't have time to fix it, but here's some style updates for Chrome that maybe you can translate to the other browsers if you like them?

```css
.jspsych-slider::-webkit-slider-runnable-track {
  appearance: none;
  -webkit-appearance: none;
  width: 100%;
  height: 8px;
  cursor: pointer;
  background: #eee;
  box-shadow: 0px 0px 0px #000000, 0px 0px 0px #0d0d0d;
  border-radius: 2px;
  border: 1px solid #aaa;
}

.jspsych-slider::-webkit-slider-thumb {
  border: 1px solid #666;
  height: 24px;
  width: 15px;
  border-radius: 5px;
  background: #ffffff;
  cursor: pointer;
  -webkit-appearance: none;
  margin-top: -9px;
}
```

**Comment by becky-gilbert** [comment]
@jodeleeuw I updated the CSS with your changes and pushed to my fork, in case you want to play around again. Just checking that this is how it should look:

**Chrome**
![chrome_slider2](https://user-images.githubusercontent.com/9041788/99005314-6bda7b00-24f5-11eb-8753-b0857528c6fe.png)

**Firefox**
![firefox_slider2](https://user-images.githubusercontent.com/9041788/99005327-7432b600-24f5-11eb-950e-437acd55ae41.PNG)

**Edge**
![edge_slider2](https://user-images.githubusercontent.com/9041788/99005345-7eed4b00-24f5-11eb-9990-9e394706ed98.PNG)

**Comment by jodeleeuw** [comment]
Yep!

Why does the edge one have to be wider?

**Comment by becky-gilbert** [comment]
Because Edge seems to limit how much the thumb can extend higher/lower than the track (or at least, I can't figure out how to make it extend any farther). Here's what it looks like in Edge with the same track and thumb height as Chrome/FF:

![edge_slider3](https://user-images.githubusercontent.com/9041788/99008360-d510bd00-24fa-11eb-93c0-782952d1ad49.PNG)

Any suggestions for a fix? I tried setting the thumb's z-index, thinking maybe it was hidden behind other elements, and that didn't work. 

So to make sure that the thumb is big enough to be easily clickable, I went with a bigger track height to increase the thumb height, and kept the thumb width the same as in other browsers. Other suggestions welcome!
